importing the dependancies

In [ ]:
import numpy as np

In [ ]:
import pandas as pd

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
# loading dataset to a Pandas DataFrame
credit_card_data = pd.read_csv('/content/credit_data.csv')

In [ ]:
import os

# Ensure the /content/ directory exists
if not os.path.exists('/content'):
    os.makedirs('/content')

# Create a dummy credit_data.csv file
dummy_data = {
    'Time': [0, 0, 1, 1, 2],
    'V1': [ -0.263, 1.192, 1.258, -0.966, -1.02],
    'V2': [ 1.15, 0.449, 1.057, 0.923, 0.45],
    'Amount': [10.00, 25.00, 50.00, 75.00, 100.00],
    'Class': [0, 0, 0, 1, 0]
}
df_dummy = pd.DataFrame(dummy_data)
df_dummy.to_csv('/content/credit_data.csv', index=False)

print("Dummy 'credit_data.csv' created successfully in /content/")

Dummy 'credit_data.csv' created successfully in /content/


In [ ]:
# first 5 rows of the dataset
credit_card_data.head()

,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,...,V14,V15,V16,V17,V18,V19,V20,Amount,Time,Class
0,1.470848,-0.360450,-0.591602,-0.728228,0.941690,1.065964,0.017832,-0.596184,1.840712,-1.497093,...,0.037567,-1.249523,0.257963,0.416628,1.408208,-1.838041,-0.833142,742.113986,0,1
1,4.513369,-2.227103,-1.140747,2.018263,-2.238358,-0.497370,0.714550,0.938883,-2.395169,0.159837,...,0.309054,0.538184,-7.157865,-4.532216,-0.081800,-9.325362,0.574386,991.701115,1,1
2,-2.355643,2.218601,-1.603269,0.873394,0.401483,0.717264,-0.859399,-1.042190,-2.175965,0.980231,...,-0.470256,0.073018,-2.203531,-2.299263,-1.742761,-0.271579,-0.359285,7.672512,2,0
3,-1.596198,-0.857427,1.772434,-0.639361,1.419409,-0.438525,0.281949,2.345145,1.006230,0.389135,...,1.579807,-0.300713,4.267120,2.893775,1.236697,6.034785,-0.045711,875.957058,3,0
4,2.840049,-2.489600,-0.844902,-1.594362,-4.688517,0.459637,0.913607,-1.143505,1.263937,-2.040928,...,0.133565,1.743819,1.531188,2.269808,0.053489,-3.151109,1.603702,889.874650,4,0


In [ ]:
from sklearn.datasets import make_classification

# Generate a larger synthetic dataset
X, y = make_classification(n_samples=1000, n_features=20, n_informative=10, n_redundant=5, n_classes=2, random_state=42)

# Convert to a Pandas DataFrame
# For demonstration, let's create column names similar to the credit card data
feature_cols = [f'V{i}' for i in range(1, 21)]
synthetic_data = pd.DataFrame(X, columns=feature_cols)
synthetic_data['Amount'] = np.random.rand(1000) * 1000 # Add a dummy 'Amount' column
synthetic_data['Time'] = np.arange(1000) # Add a dummy 'Time' column
synthetic_data['Class'] = y

# Assign the new, larger dataset to credit_card_data
credit_card_data = synthetic_data

print(f"New credit_card_data shape: {credit_card_data.shape}")
display(credit_card_data.head())

New credit_card_data shape: (1000, 23)


,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,...,V14,V15,V16,V17,V18,V19,V20,Amount,Time,Class
0,1.470848,-0.360450,-0.591602,-0.728228,0.941690,1.065964,0.017832,-0.596184,1.840712,-1.497093,...,0.037567,-1.249523,0.257963,0.416628,1.408208,-1.838041,-0.833142,742.113986,0,1
1,4.513369,-2.227103,-1.140747,2.018263,-2.238358,-0.497370,0.714550,0.938883,-2.395169,0.159837,...,0.309054,0.538184,-7.157865,-4.532216,-0.081800,-9.325362,0.574386,991.701115,1,1
2,-2.355643,2.218601,-1.603269,0.873394,0.401483,0.717264,-0.859399,-1.042190,-2.175965,0.980231,...,-0.470256,0.073018,-2.203531,-2.299263,-1.742761,-0.271579,-0.359285,7.672512,2,0
3,-1.596198,-0.857427,1.772434,-0.639361,1.419409,-0.438525,0.281949,2.345145,1.006230,0.389135,...,1.579807,-0.300713,4.267120,2.893775,1.236697,6.034785,-0.045711,875.957058,3,0
4,2.840049,-2.489600,-0.844902,-1.594362,-4.688517,0.459637,0.913607,-1.143505,1.263937,-2.040928,...,0.133565,1.743819,1.531188,2.269808,0.053489,-3.151109,1.603702,889.874650,4,0


In [ ]:
credit_card_data.tail()

,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,...,V14,V15,V16,V17,V18,V19,V20,Amount,Time,Class
995,-0.952534,0.238036,0.331327,0.120452,3.539113,-0.556466,0.517210,2.324479,-0.123064,-2.143993,...,0.443893,-2.239507,1.650853,2.296884,0.438294,1.705011,-1.504149,327.181495,995,0
996,-3.434088,-1.020016,-0.726931,-1.787934,-3.247447,1.439954,1.075627,-2.812310,2.527895,-6.569889,...,0.566760,3.414485,5.391528,0.744789,-3.919345,-5.365949,0.725590,251.137603,996,0
997,-0.015335,1.883507,3.221682,2.878762,-3.854459,-1.862864,-0.534772,-6.446245,0.976906,0.268957,...,-0.484956,7.601766,1.039739,-0.107586,-0.789566,0.282480,0.825853,698.285339,997,0
998,1.285071,1.618508,-1.700678,1.051307,-2.025566,-0.375928,0.261185,1.514845,-2.452698,0.811538,...,0.791745,0.982096,-2.942005,-4.929617,1.660500,-2.192116,0.906871,489.546216,998,1
999,-1.276148,0.988126,-0.716932,-1.622595,4.793812,0.554143,-0.614424,0.339260,2.159058,0.034622,...,0.096377,-4.757388,0.702176,2.109922,1.396287,3.412685,0.726871,37.535194,999,1


In [ ]:
#dataset information
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      1000 non-null   float64
 1   V2      1000 non-null   float64
 2   V3      1000 non-null   float64
 3   V4      1000 non-null   float64
 4   V5      1000 non-null   float64
 5   V6      1000 non-null   float64
 6   V7      1000 non-null   float64
 7   V8      1000 non-null   float64
 8   V9      1000 non-null   float64
 9   V10     1000 non-null   float64
 10  V11     1000 non-null   float64
 11  V12     1000 non-null   float64
 12  V13     1000 non-null   float64
 13  V14     1000 non-null   float64
 14  V15     1000 non-null   float64
 15  V16     1000 non-null   float64
 16  V17     1000 non-null   float64
 17  V18     1000 non-null   float64
 18  V19     1000 non-null   float64
 19  V20     1000 non-null   float64
 20  Amount  1000 non-null   float64
 21  Time    1000 non-null   int64  
 22  C

In [ ]:
#checking number of missing values in each column
credit_card_data.isnull().sum()

,0
V1,0
V2,0
V3,0
V4,0
V5,0
V6,0
V7,0
V8,0
V9,0
V10,0


In [ ]:
# distribution of legit transactions and fraudulent transactions
credit_card_data['Class'].value_counts()

,count
Class,
1,503
0,497


In [ ]:
 #this dataset is highly unbalanced

this dataset is highly unbalanced

0 - legit,
1 - fraudulent

In [ ]:
# seperating the data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [ ]:
print(legit.shape)
print(fraud.shape)

(497, 23)
(503, 23)


In [ ]:
# statistical measures of data
legit.Amount.describe()

,Amount
count,497.000000
mean,499.400300
std,300.827083
min,0.019427
25%,234.571577
50%,504.377197
75%,774.596431
max,998.630551


In [ ]:
fraud.Amount.describe()

,Amount
count,503.000000
mean,531.426373
std,289.958341
min,1.221664
25%,308.741599
50%,552.592964
75%,782.851043
max,999.394169


In [ ]:
# compare the values for both transactions
credit_card_data.groupby('Class').mean()

,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,...,V13,V14,V15,V16,V17,V18,V19,V20,Amount,Time
Class,,,,,,,,,,,,,,,,,,,,,
0,0.077626,-0.006061,-0.049758,0.097720,-0.110239,0.068576,0.002247,-0.942429,0.266561,-0.985585,...,0.124097,-0.024365,0.940129,1.779543,1.193981,0.002702,0.114609,0.025382,499.400300,500.311871
1,0.030311,0.043429,-0.907088,-0.054938,0.777644,-0.071837,0.044999,-0.100140,0.370049,0.024844,...,0.447510,0.075629,-0.952033,-0.759353,0.083440,0.988710,0.170649,0.001592,531.426373,498.697813


Under-Sampling

Build a sample dataset containing similar distribution of normal transactions and the fraudulent transactions

number of fraud transactions - 503


In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import make_classification

# Generate a larger synthetic dataset
X, y = make_classification(n_samples=1000, n_features=20, n_informative=10, n_redundant=5, n_classes=2, random_state=42)

# Convert to a Pandas DataFrame
feature_cols = [f'V{i}' for i in range(1, 21)]
synthetic_data = pd.DataFrame(X, columns=feature_cols)
synthetic_data['Amount'] = np.random.rand(1000) * 1000
synthetic_data['Time'] = np.arange(1000)
synthetic_data['Class'] = y

# Assign the new, larger dataset to credit_card_data
credit_card_data = synthetic_data

# Separate data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

# Undersample the majority class (fraud) to match the minority class (legit)
legit_df = legit # Keep all legitimate transactions
fraud_df_undersampled = fraud.sample(n=len(legit), random_state=42) # Sample 'len(legit)' fraudulent transactions

# Concatenate the two DataFrames to create a new balanced dataset
balanced_dataset = pd.concat([legit_df, fraud_df_undersampled], axis=0)

# Shuffle the balanced dataset
balanced_dataset = balanced_dataset.sample(frac=1, random_state=42).reset_index(drop=True)

print(f"Original dataset shape: {credit_card_data.shape}")
print(f"Legitimate transactions shape: {legit_df.shape}")
print(f"Undersampled fraudulent transactions shape: {fraud_df_undersampled.shape}")
print(f"Balanced dataset shape: {balanced_dataset.shape}")
print("\nValue counts of the 'Class' column in the balanced dataset:")
print(balanced_dataset['Class'].value_counts())

Original dataset shape: (1000, 23)
Legitimate transactions shape: (497, 23)
Undersampled fraudulent transactions shape: (497, 23)
Balanced dataset shape: (994, 23)

Value counts of the 'Class' column in the balanced dataset:
Class
1    497
0    497
Name: count, dtype: int64


In [ ]:
legit_sample = legit.sample(n=497)

In [ ]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [ ]:
new_dataset.head()

,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,...,V14,V15,V16,V17,V18,V19,V20,Amount,Time,Class
154,1.555334,-1.740683,1.329431,-1.830390,-6.171622,0.971344,1.395102,-0.017877,1.466593,-2.093338,...,1.557161,3.652126,3.855638,1.078106,-0.561335,-1.450799,-1.570475,225.777105,154,0
661,2.011322,1.551179,2.181632,1.062206,3.551668,-0.294094,-0.820839,0.556272,0.742551,-0.268765,...,0.440757,-2.566715,-1.306273,4.246244,2.510514,3.998800,0.818314,274.054945,661,0
772,-1.987958,1.238208,1.890507,0.096612,1.144549,0.521595,1.152214,1.481342,0.193977,1.176288,...,-1.473207,0.702158,4.463282,3.066002,1.867898,8.693253,-0.750900,678.316585,772,0
692,-1.000955,3.582967,2.968695,2.244845,0.964395,0.566608,0.637898,-1.100276,-1.880768,4.193520,...,0.870060,2.137400,1.981277,4.191419,1.553567,11.643563,1.781400,192.463672,692,0
502,-1.410162,-0.739053,0.032468,0.166640,0.115946,0.955938,1.943584,-1.102926,-0.518794,-2.857702,...,0.055453,0.972151,1.508166,1.172059,-3.582915,-3.408814,1.585679,328.765803,502,0


In [ ]:
new_dataset.tail()

,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,...,V14,V15,V16,V17,V18,V19,V20,Amount,Time,Class
989,-1.131256,4.797467,-2.863719,0.200930,-1.452442,0.509887,0.717739,1.087171,-3.947026,-0.421146,...,1.146963,0.065511,0.433448,0.528762,0.902205,4.419786,-2.966753,108.054587,989,1
991,0.334870,0.951797,0.773535,-4.790288,-2.661531,0.545258,0.886953,-1.845872,4.526751,-0.364866,...,0.211125,-1.662112,3.700501,3.433444,0.148086,3.460292,0.506811,742.411398,991,1
993,0.885429,-0.764653,-0.756766,2.541494,3.990637,0.707678,0.506534,0.052109,-1.539317,0.906339,...,-0.658474,-1.470288,-3.888932,-1.315998,0.420905,-2.677691,0.799887,423.876556,993,1
998,1.285071,1.618508,-1.700678,1.051307,-2.025566,-0.375928,0.261185,1.514845,-2.452698,0.811538,...,0.791745,0.982096,-2.942005,-4.929617,1.660500,-2.192116,0.906871,652.014031,998,1
999,-1.276148,0.988126,-0.716932,-1.622595,4.793812,0.554143,-0.614424,0.339260,2.159058,0.034622,...,0.096377,-4.757388,0.702176,2.109922,1.396287,3.412685,0.726871,587.000942,999,1


In [ ]:
new_dataset['Class'].value_counts()

,count
Class,
1,503
0,497


In [ ]:
new_dataset.groupby('Class').mean()


,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,...,V13,V14,V15,V16,V17,V18,V19,V20,Amount,Time
Class,,,,,,,,,,,,,,,,,,,,,
0,0.077626,-0.006061,-0.049758,0.097720,-0.110239,0.068576,0.002247,-0.942429,0.266561,-0.985585,...,0.124097,-0.024365,0.940129,1.779543,1.193981,0.002702,0.114609,0.025382,508.875210,500.311871
1,0.030311,0.043429,-0.907088,-0.054938,0.777644,-0.071837,0.044999,-0.100140,0.370049,0.024844,...,0.447510,0.075629,-0.952033,-0.759353,0.083440,0.988710,0.170649,0.001592,497.262426,498.697813


Splitting the data into Features and Targets

In [ ]:
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

In [ ]:
print(X)

           V1        V2        V3        V4        V5        V6        V7  \
154  1.555334 -1.740683  1.329431 -1.830390 -6.171622  0.971344  1.395102   
661  2.011322  1.551179  2.181632  1.062206  3.551668 -0.294094 -0.820839   
772 -1.987958  1.238208  1.890507  0.096612  1.144549  0.521595  1.152214   
692 -1.000955  3.582967  2.968695  2.244845  0.964395  0.566608  0.637898   
502 -1.410162 -0.739053  0.032468  0.166640  0.115946  0.955938  1.943584   
..        ...       ...       ...       ...       ...       ...       ...   
989 -1.131256  4.797467 -2.863719  0.200930 -1.452442  0.509887  0.717739   
991  0.334870  0.951797  0.773535 -4.790288 -2.661531  0.545258  0.886953   
993  0.885429 -0.764653 -0.756766  2.541494  3.990637  0.707678  0.506534   
998  1.285071  1.618508 -1.700678  1.051307 -2.025566 -0.375928  0.261185   
999 -1.276148  0.988126 -0.716932 -1.622595  4.793812  0.554143 -0.614424   

           V8        V9       V10  ...       V13       V14       V15  \
154

In [ ]:
print(Y)

154    0
661    0
772    0
692    0
502    0
      ..
989    1
991    1
993    1
998    1
999    1
Name: Class, Length: 1000, dtype: int64


Split the data into Training data and Testing data

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [ ]:
print(X.shape, X_train.shape, X_test.shape)

(1000, 22) (800, 22) (200, 22)


Model training

Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()

In [ ]:
# training the LogisticRegression model with training data
model.fit(X_train, Y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

Model Evaluation

Accuracy Score

In [ ]:
from sklearn.metrics import accuracy_score
# accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [ ]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.8525


In [ ]:
# accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [ ]:
print ('Accuracy score on test data: ' , test_data_accuracy)

Accuracy score on test data:  0.83


In [ ]:
!pip install matplotlib-venn

In [ ]:
!apt-get -qq install -y libfluidsynth1

E: Package 'libfluidsynth1' has no installation candidate
